business/services.go

```diff
@@ -252,11 +252,67 @@ func (in *SvcService) buildKubernetesServices(svcs []core_v1.Service, pods []cor
 	return services
 }
 
+// The istiod registry doesn't have a explicit flag when a service is deployed in a different control plane.
+// The only way to identify it is to check that the service has an address in the current cluster.
+// To avoid side effects, Kiali will process only services that belongs to the current cluster.
+// This should be revisited on more multi-cluster deployments scenarios.
+//     "hostname": "test-svc.evil.svc.cluster.local",
+//    "clusterVIPs": {
+//      "Addresses": {
+//        "istio-west": [
+//          "0.0.0.0"
+//        ]
+//      }
+//    },
+func (in *SvcService) getClusterId() string {
+	// By default Istio uses "Kubernetes" as clusterId for single control planes scenarios.
+	// This clusterId is propagated into the Istio Registry and we need it to filter services in multi-cluster scenarios.
+	// I.e.:
+	//    "clusterVIPs": {
+	//      "Addresses": {
+	//        "Kubernetes": [
+	//          "10.217.4.189"
+	//        ]
+	//      }
+	//    }
+	clusterId := DefaultClusterID
+	// Protection on tests
+	if in.businessLayer != nil {
+		if cluster, err := in.businessLayer.Mesh.ResolveKialiControlPlaneCluster(nil); err == nil {
```

Comment:
I wonder if this works for single cluster.
I remember this can return (nil, nil).

Let me try...

Reply:
OK. Looks like it works as long as Istio and Kiali are both installed on the same cluster. If it happens that Istio is not found, the error I see is `Could not fetch services list: unable to find any healthy Pilot instance`. So the services list is protected from a `panic`.

So, this is OK.